### PR TITLE
MM-12844/MM-13001 Add interactive dialogs

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
First two commits refactor components into two new widgets: TextSetting and AutocompleteSelector.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-12844

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (https://github.com/mattermost/mattermost-server/pull/9816)
- [x] Has redux changes (https://github.com/mattermost/mattermost-redux/pull/710)
- [x] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates

### CODE REVIEW

**Comment by jwilander** [comment]
@jasonblais should be ready for PM review. I've added slash commands `/dialog` that will open the dialog and both an interactive message button and menu that should do the same.

I'll need a bit of help from @asaadmahmood with some the UI bits but everything should be working functionally.

If we need new a test server let me know as I'll need to set up the integration again

**Comment by jasonblais** [comment]
@jwilander Can you help share (here or via DM) the JSON payload you used for this integration? I'll just use it as a reference for docs.

**Comment by lindy65** [comment]
Thanks @jwilander @jasonblais 

1. Noted that UI work still needs to be done by Asaad so will test UI separately once done.

2. Is there meant to be a `Submitted` confirmation to the right of `Action button - "press me"` (similar to `Submitted` to the right of `Option` menu)? (Although an ephemeral message is posted confirming "you pressed the button".)

image

image

3. Tested using the slash command `/dialog` and this works as expected from any channel, DM and GM, posting an ephemeral message to confirm your command (the same as using any other slash command would)

4. A question relating to 3) and using the `/dialog` slash command from any channel
image

   - From the example above, the `You used a slash command` is obviously configurable to give a more explanatory confirmation message (as displayed by `/expand` and `dnd` ephemeral messages (but it is posted in the channel before the dialog form is even submitted). I was just wondering if there is any kind of feedback to the user if submitting the correctly completed (i.e. no missing required data) dialog form fails?

5. From the [spec](https://docs.google.com/document/d/1uAv1ajzEWJErccGp77hD0i2PY4e9nLsZ_oM77iTXIJU/edit#heading=h.233hb1bdwh3i), it seems I should be able to type 3000 characters in the `Display Name Long Text Area` but I can only type 100 before input is disabled.

6. The required fields, `User Selector` and `Option Selector` don't seem to validate data if I insert something that's not in the list and the dialog form is submitted without error.

image

7. Although `Channel Selector` is an optional field, data doesn't seem to be validated either if I insert something that's not in the list and the dialog form is submitted without error.

image

8. Also from the [spec](https://docs.google.com/document/d/1uAv1ajzEWJErccGp77hD0i2PY4e9nLsZ_oM77iTXIJU/edit#bookmark=id.gd4e6vzi5vgr) I'm not sure what "clicks away" means.  If the dialog form is open and I click just outside it, it closes (as if I'd clicked on the `X` or `Cancel`).  Is this expected or should the dialog stay open?

9. Tested
   a. trying to submit the dialog with no data - returns correct error messages for required fields
   b. tried an invalid email address - returns error message `Must be a valid email address.`
   c. cancelling out of the dialog form - works as expected
   d. `dialog` slash command in RHS - works as expected
   e. `action button` and `attachment text` in RHS - works as expected

**Comment by jasonblais** [comment]
Thanks Lindy! This is great.

@jwilander Looks like points 6-8 are potential bugs.

I also tested on mobile browser and didn't find additional non-UI issues from what Lindy mentioned. The different mobile keyboard layouts based on the subtype of the field worked for me (tested email, text, number)

UX ticket is queued for Asaad to review UI: https://mattermost.atlassian.net/browse/MM-13089

----

@lindy65 

2 - No, buttons don't currently have the `Submitted` label. I'm also discussing with Asaad if this label adds enough value, given the integration will likely respond back with an ephemeral message and there are a few issues that currently exist with that label.

4 - This will be handled by the integration itself - for instance, Joram could add a custom ephemeral message to his integration when the dialog was cancelled or when the info was submitted.

5 - I looked at the JSON output Joram shared with me and he used the following fields for `Display Name Long Text Area`:
 - "min_length":5,
 - "max_length":100,

These are optional fields that can be used - given the max length of 100, you're seeing the truncation at 100 characters.

6 /7 - Will defer to @jwilander 

8 - The dialog should only close if the user clicks on 'X' or 'Cancel'. Clicking outside of the dialog (e.g. in the LHS) shouldn't close it, so this might be a bug @jwilander

**Comment by jwilander** [comment]
6/7 - This is a bit hard to validate for users and channels (how do we know if what was typed is a user or not). @jasonblais maybe we can talk about this some more but we might want to think about forcing the user to select from the dropdown and clearing any text from the field if they do not

8 - Will fix

**Comment by lindy65** [comment]
Thanks for the responses to 2) 4) and 5) @jasonblais :)

Apologies for missing to test on mobile browser!

@jwilander Agree with you on
> forcing the user to select from the dropdown and clearing any text from the field if they do not

for 6) and 7)

and will wait for new test server to re-test 8)

**Comment by jasonblais** [comment]
6/7 - discussed with Joram and we agreed to go with the proposed solution where the text field is cleared if no option is selected. The same will be applied for interactive menus part of message attachments.

**Comment by jwilander** [comment]
Updates made for 6/7/8, I'm going to wait to update the spinmint until the UX changes are ready so I don't have to set up the integration again twice (let me know if you want to test it before though, I can update it sooner)

**Comment by jasonblais** [comment]
@jwilander That sounds good. Testing 6 - 8 should be pretty quick so we can do that after UX changes are done.

File: actions/integration_actions.jsx [hunk 1]
```diff
@@ -110,3 +115,22 @@ export function getYoutubeVideoInfo(googleKey, videoId, success, error) {
...
+
+    previousTriggerId = currentTriggerId;
+
+    const dialog = state.entities.integrations.dialog || {};
+    if (dialog.trigger_id !== currentTriggerId) {
+        return;
+    }
+
+    store.dispatch(openModal({modalId: ModalIdentifiers.INTERACTIVE_DIALOG, dialogType: InteractiveDialog}));
+});
```
[the remark this answers is not in the document]
Comment (jwilander) [reply actions/integration_actions.jsx]: Specifically, this is to handle a race where we receive the open dialog WS event before the post action/slash command/whatever returns us a trigger ID to watch for. As for why it's globally scoped there really wasn't another good way to go about it IMO. I could have created a dummy component to include on the root that renders null and uses its props to fire off this action but to me that was more confusing than just having a global listener.

[the remark this answers is not in the document]
Comment (grundleborg) [reply actions/integration_actions.jsx]: Thanks for the explanation - makes perfect sense to me.

**Comment by jwilander** [comment]
I'll re-up the server once the build completes

**Comment by jwilander** [comment]
@jasonblais integration all set up and ready to go

**Comment by jasonblais** [comment]
@jwilander Have tested 6 - 8, these are fixed.

Tested on Mac/Chrome in desktop and mobile view, as well as on iOS/Safari. Will test RN app next.

Observations:

1 - The message menu didn't trigger a dialog - is that a mistake when setting up the integration, or an issue on the PR?
2 - @asaadmahmood On mobile view, two things
  - the header bleeds into the text fields. This also reproduces on webapp, but it's more noticeable on mobile view. Should we add a separator between the header and the body?
  - I'm unable to scroll all the way to the bottom of the dialog because of the Cancel/Submit buttons

image

**Comment by jwilander** [comment]
1 - Sorry that was a bug in the test integration, should be fixed now @jasonblais

**Comment by asaadmahmood** [comment]
@jasonblais Pushed.

**Comment by asaadmahmood** [comment]
localhost_8065_mattermost_channels_town-square
localhost_8065_mattermost_channels_town-square iphone 5_se

**Comment by jasonblais** [comment]
For the first image (webapp view), does the header bleed into the text areas if you scroll down like in the attached below? Or does the separator also appear in webapp view?

image

**Comment by jasonblais** [comment]
@jwilander I think Asaad is out for the day. This last remaining UI question isn't a blocker so let's merge this. I'll follow up with Asaad separately next week.

**Pull request merged**